### PR TITLE
feat(ref-imp): #766 - updated error handling for file downloading and validation

### PR DESCRIPTION
**Pull request by thehenrytsai**
Changed the download and validation flow of Sidetree files to match decisions made in WG.

### CODE REVIEW

File: lib/core/versions/latest/BatchWriter.ts [hunk 1]
```diff
@@ -109,6 +102,43 @@ export default class BatchWriter implements IBatchWriter {
...
+  }
+
+  /**
+   * Create and write provisional index file if needed.
+   * @returns CASE URI of the chunk file. `undefined` if there is no need to create and write the file.
+   */
+  private async createAndWriteProvisionalIndexFileIfNeeded (
+    chunkFileUri: string | undefined, provisionalProofFileHash: string | undefined, updateOperations: UpdateOperation[]
+  ): Promise<string | undefined> {
+    // If `chunkFileHash` is `undefined` it means there are only deactivates, and a batch with only deactivates does not reference a provisional index file.
```
Comment (isaacJChen) [inline lib/core/versions/latest/BatchWriter.ts:130]:
```suggestion
    // If `chunkFileUri` is `undefined` it means there are only deactivates, and a batch with only deactivates does not reference a provisional index file.
```

File: lib/core/versions/latest/BatchWriter.ts [hunk 2]
```diff
@@ -109,6 +102,43 @@ export default class BatchWriter implements IBatchWriter {
     await this.operationQueue.dequeue(queuedOperations.length);
   }
 
+  /**
+   * Create and write chunk file if needed.
+   * @returns CASE URI of the chunk file. `undefined` if there is no need to create and write the file.
```
Comment (isaacJChen) [inline lib/core/versions/latest/BatchWriter.ts:107]:
```suggestion
   * @returns CAS URI of the chunk file. `undefined` if there is no need to create and write the file.
```

File: lib/core/versions/latest/BatchWriter.ts [hunk 3]
```diff
@@ -109,6 +102,43 @@ export default class BatchWriter implements IBatchWriter {
...
+
+    const chunkFileUri = await this.cas.write(chunkFileBuffer);
+    console.info(LogColor.lightBlue(`Wrote chunk file ${LogColor.green(chunkFileUri)} to content addressable store.`));
+
+    return chunkFileUri;
+  }
+
+  /**
+   * Create and write provisional index file if needed.
+   * @returns CASE URI of the chunk file. `undefined` if there is no need to create and write the file.
```
Comment (isaacJChen) [inline lib/core/versions/latest/BatchWriter.ts:125]:
```suggestion
   * @returns CAS URI of the chunk file. `undefined` if there is no need to create and write the file.
```

File: lib/core/versions/latest/ChunkFile.ts [hunk 4]
```diff
@@ -60,13 +60,19 @@ export default class ChunkFile {
 
   /**
    * Creates chunk file buffer.
+   * @returns Chunk file buffer. Returns `undefined` arrays passed in contains no operations.
```
Comment (isaacJChen) [inline lib/core/versions/latest/ChunkFile.ts:63]:
```suggestion
   * @returns Chunk file buffer. Returns `undefined` if arrays passed in contains no operations.
```

File: lib/core/versions/latest/BatchWriter.ts (hunk 1 above)
Comment (thehenrytsai) [reply lib/core/versions/latest/BatchWriter.ts:130]: Thanks, replaced all occurrences in code base.

File: lib/core/versions/latest/BatchWriter.ts (hunk 3 above)
Comment (thehenrytsai) [reply lib/core/versions/latest/BatchWriter.ts:125]: Thanks!

File: lib/core/versions/latest/BatchWriter.ts (hunk 2 above)
Comment (thehenrytsai) [reply lib/core/versions/latest/BatchWriter.ts:107]: Fixed.

**Review by isaacJChen — APPROVED** [review]

**Pull request merged**